### ISSUE TITLE
Incorrect assembly generated for __attribute__((uncached)) access

### ISSUE THREAD

**Issue by petrokarashchenko** [foss-for-synopsys-dwc-arc-processors/toolchain#220]
Compiler "GCC: (ARCompact/ARCv2 ISA elf32 toolchain 2019.09-rc2) 9.2.1 20191002"
C code:
```
#define RegWr(a,v)     (*(volatile __attribute__((uncached)) uint32_t *)(a)=(v))

void my_func(int arg1, uint32_t arg2) {
    uint32_t reg = 0;

    switch(arg1) {
    case 0:
        RegWr(0x44100, arg2);
        reg |= 0x02;
        RegWr(0x668, reg);
        break;
    case 1:
        RegWr(0x44380, arg2);
        reg |= 0x04;
        RegWr(0x668, reg);
        break;
    case 2:
        RegWr(0x45480, arg2);
        reg |= 0x08;
        RegWr(0x668, reg);
        break;
    default:
        break;
    }
}
```
Generated assembly:
```
20004014 <my_func>:
20004014:	0821 0050           	breq	r0,0x1,32	;20004034 <my_func+0x20>
20004018:	0829 00d0           	breq	r0,0x3,40	;20004040 <my_func+0x2c>
2000401c:	780b                	tst_s	r0,r0
2000401e:	7de0                	jne_s	[blink]
20004020:	1e00 7060 0004 4100 	st.di	r1,[0x44100]
20004028:	d802                	mov_s	r0,0x2
2000402a:	1e00 7000 0000 0668 	st	r0,[0x668]
20004032:	7ee0                	j_s	[blink]
20004034:	1e00 7060 0004 4380 	st.di	r1,[0x44380]
2000403c:	d804                	mov_s	r0,0x4
2000403e:	f1f7                	b_s	-18	;2000402a <my_func+0x16>
20004040:	1e00 7060 0004 5480 	st.di	r1,[0x45480]
20004048:	d808                	mov_s	r0,0x8
2000404a:	f1f1                	b_s	-30	;2000402a <my_func+0x16>
```

The problem is that it is expected `st.di r0,[0x668]`, but actually GCC generates `st r0,[0x668]`

**Comment by petrokarashchenko** [comment]
Is it possible to fix this is in GNU Toolchain for ARC Processors, 2019.09 release?

**Comment by petrokarashchenko** [comment]
@fbedard could you please take a look

**Comment by petrokarashchenko** [comment]
CC: @claziss

**Comment by claziss** [comment]
Hi,

Strangely enough, I am getting what it is expected, using `-mcpu=arcem -O1` options for the compiler:
```asm
00000000 <my_func>:
   0:	0821 0050           	breq.nt	r0,0x1,32	;20 <my_func+0x20>
   4:	0831 0090           	breq.nt	r0,0x2,48	;34 <my_func+0x34>
   8:	780b                	tst_s	r0,r0
   a:	7de0                	jne_s	[blink]
   c:	1e00 7060 0004 4100 	st.di	r1,[0x44100]
  14:	1e00 70a1 0000 0668 	st.di	2,[0x668]
  1c:	7ee0                	j_s	[blink]
  1e:	78e0                	nop_s
  20:	1e00 7060 0004 4380 	st.di	r1,[0x44380]
  28:	1e00 7121 0000 0668 	st.di	4,[0x668]
  30:	7ee0                	j_s	[blink]
  32:	78e0                	nop_s
  34:	1e00 7060 0004 5480 	st.di	r1,[0x45480]
  3c:	1e00 7221 0000 0668 	st.di	8,[0x668]
  44:	7ee0                	j_s	[blink]
  46:	78e0                	nop_s
```
Please, can u tell me exactly what compiler options do u use.

**Comment by petrokarashchenko** [comment]
Hi @claziss,
 
Compiler flags are next:
```
CFLAGS   = -mcpu=arc600 -mtune=arc600 -mbig-endian -mmul64 -mbarrel-shifter
CFLAGS  += -ffunction-sections -fdata-sections -fno-common -fno-strict-aliasing -c -Os
CFLAGS  += -Wa,-adhlns="$@.lst"
```

**Comment by claziss** [comment]
Go it. Thanks, I will come with a solution asap.

**Comment by claziss** [comment]
The issue seems to be one of the GCC passes (i.e. `jump2`) which decides to drop the memory attributes leading to the observed problem. As a work around is to use `-fno-crossjumping` compiler option until I properly fix it.

**Comment by petrokarashchenko** [comment]
Thank you @claziss for your support. I will try it

**Comment by petrokarashchenko** [comment]
I confirm that workaround works

**Comment by abrodkin** [comment]
@claziss I guess we need to bump a milestone to `2020.03` as proper fix was not done yet and then it's not in 2019.09, right?

**Comment by claziss** [comment]
This issue should be fixed here:
https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/commit/2080016aa654d7d4b8dc4fddc1e151d5ffa35d54

**Comment by petrokarashchenko** [comment]
All my issues related to `uncached` attribute are fixed using latest commits from `2020.03` gcc branch

**Comment by claziss** [comment]
@petrokarashchenko I've rebased/merged/fixed the 2030.03. Please refer to the latest. I'll close this issue, and keep only one.

**Comment by abrodkin** [comment]
Fixed by https://github.com/foss-for-synopsys-dwc-arc-processors/gcc/commit/fc5aa9c7be10d21206537161f2245c097e06f456 in `arc-2020.03`.
